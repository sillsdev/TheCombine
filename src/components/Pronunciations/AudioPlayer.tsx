import {
  makeStyles,
  Theme,
  createStyles,
  Tooltip,
  IconButton
} from "@material-ui/core";
import { green } from "@material-ui/core/colors";
import PlayArrow from "@material-ui/icons/PlayArrow";
import React from "react";
import { Translate } from "react-localize-redux";
import { Stop } from "@material-ui/icons";

export interface PlayerProps {
  pronunciationUrl: string;
  isPlaying?: boolean;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    button: {
      margin: theme.spacing(1)
    },
    icon: {
      color: green[800]
    }
  })
);

export default function AudioPlayer(props: PlayerProps) {
  const [playing, setPlaying] = React.useState<boolean>(false);
  const [audio] = React.useState<HTMLAudioElement>(
    new Audio(props.pronunciationUrl)
  );

  const classes = useStyles();
  // const audio = new Audio(props.pronunciationUrl);
  // audio.crossOrigin = "annonymous";

  let togglePlay = () => {
    if (!playing) {
      audio.play();
      setPlaying(true);
      audio.addEventListener("ended", () => setPlaying(false));
    } else {
      audio.pause();
      setPlaying(false);
      audio.currentTime = 0;
    }
  };

  return (
    <Tooltip title={<Translate id="pronunciations.playTooltip" />}>
      <IconButton
        onClick={togglePlay}
        className={classes.button}
        aria-label="play"
      >
        {playing ? (
          <Stop className={classes.icon} />
        ) : (
          <PlayArrow className={classes.icon} color="primary" />
        )}
      </IconButton>
    </Tooltip>
  );
}
