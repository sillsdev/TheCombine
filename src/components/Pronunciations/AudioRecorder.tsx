import {
  makeStyles,
  Theme,
  createStyles,
  IconButton,
  Tooltip
} from "@material-ui/core";
import { red } from "@material-ui/core/colors";
import FiberManualRecord from "@material-ui/icons/FiberManualRecord";
import React, { useState } from "react";
import * as Backend from "../../backend";
import { Recorder } from "./Recorder";
import { Translate } from "react-localize-redux";

export interface RecorderProps {
  wordId: string;
  recordingFinished?: (oldId: string, newId: string) => void;
}

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    button: {
      margin: theme.spacing(1)
    },
    icon: {
      color: red[800]
    }
  })
);

function getFileNameForWord(wordId: string): string {
  var fourCharParts = wordId.match(/.{1,6}/g);
  var compressed =
    fourCharParts === null
      ? ["unknownword"]
      : fourCharParts.map(i => Number("0x" + i).toString(36));
  return compressed.join("") + "_" + new Date().getTime().toString(36);
}

export default function AudioRecorder(props: RecorderProps) {
  const [isRecording, setIsRecording] = useState(false);
  const classes = useStyles();
  const audio = new Recorder();

  return (
    <Tooltip title={<Translate id="pronunciations.recordTooltip" />}>
      <IconButton
        onMouseDown={() => {
          audio.startRecording();
          setIsRecording(true);
        }}
        onMouseUp={() => {
          if (isRecording === true) {
            audio
              .stopRecording()
              .then((audioUrl: string) => {
                const blob = audio.getBlob();
                const fileName = getFileNameForWord(props.wordId);
                const file = new File([blob], fileName, {
                  type: blob.type,
                  lastModified: Date.now()
                });
                Backend.uploadAudio(props.wordId, file).then(newWordId => {
                  audio.clearData();
                  if (props.recordingFinished) {
                    props.recordingFinished(props.wordId, newWordId);
                  }
                });
              })
              .catch(() => {
                console.log("Error recording, probably no mic access");
                // <Translate id="pronunciations.noMicAccess" />;
                // TODO: Show alert dialog here
              });
          }
        }}
        className={classes.button}
        aria-label="record"
      >
        <FiberManualRecord className={classes.icon} />
      </IconButton>
    </Tooltip>
  );
}
