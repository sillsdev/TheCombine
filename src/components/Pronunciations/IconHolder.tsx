import React from "react";
import FiberManualRecord from "@material-ui/icons/FiberManualRecord";
import { makeStyles, IconButton } from "@material-ui/core";
import { red } from "@material-ui/core/colors";
import { useSelector, useDispatch } from "react-redux";
import { updateRecordingStatus } from "../../goals/ReviewEntries/ReviewEntriesComponent/ReviewEntriesActions";

export interface IconHolderProps {
  wordId: string;
  safeStartRecording: () => void;
  safeStopRecording: () => void;
}

export interface IconHolderState {
  isRecording: boolean;
}

export default function IconHolder(props: IconHolderProps) {
  const isRecording = useSelector(
    (state: any) => state.reviewEntriesState.isRecording
  );
  const wordBeingRecorded = useSelector(
    (state: any) => state.reviewEntriesState.wordBeingRecorded
  );
  const dispatch = useDispatch();

  const useStyles = makeStyles((theme) => ({
    button: {
      margin: theme.spacing(1),
    },
    iconPress: {
      color: red[900],
    },
    iconRelease: {
      color: red[500],
    },
  }));

  const classes = useStyles();

  function toggleIsRecordingToTrue(
    event: Event | React.TouchEvent | React.MouseEvent
  ) {
    event.preventDefault();
    dispatch(updateRecordingStatus(true, props.wordId));
    props.safeStartRecording();
  }
  function toggleIsRecordingToFalse(
    event: Event | React.TouchEvent | React.MouseEvent
  ) {
    event.preventDefault();
    props.safeStopRecording();
    dispatch(updateRecordingStatus(false, undefined));
  }

  return (
    <IconButton
      onMouseDown={toggleIsRecordingToTrue}
      onTouchStart={toggleIsRecordingToTrue}
      onMouseUp={toggleIsRecordingToFalse}
      onTouchEnd={toggleIsRecordingToFalse}
      className={classes.button}
      aria-label="record"
    >
      <FiberManualRecord
        className={
          isRecording && props.wordId === wordBeingRecorded
            ? classes.iconPress
            : classes.iconRelease
        }
      />
    </IconButton>
  );
}
