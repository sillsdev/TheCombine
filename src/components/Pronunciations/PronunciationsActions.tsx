export enum PronunciationsStatus {
  Default = "DEFAULT",
  Playing = "PLAYING",
  Recording = "RECORDING",
}

export interface PronunciationsAction {
  type: PronunciationsStatus;
  payload?: string;
}

export function playing(payload: string): PronunciationsAction {
  return {
    type: PronunciationsStatus.Playing,
    payload,
  };
}

export function recording(payload: string): PronunciationsAction {
  return {
    type: PronunciationsStatus.Recording,
    payload,
  };
}

export function reset(): PronunciationsAction {
  return { type: PronunciationsStatus.Default };
}
