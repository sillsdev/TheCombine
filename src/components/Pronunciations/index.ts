import { connect } from "react-redux";
import { StoreState } from "../../types";
import { getTranslate } from "react-localize-redux";
import Pronunciations from "./PronunciationsComponent";

function mapStateToProps(state: StoreState) {
  return {
    translate: getTranslate(state.localize)
  };
}

export default connect(
  mapStateToProps,
  null
)(Pronunciations);
