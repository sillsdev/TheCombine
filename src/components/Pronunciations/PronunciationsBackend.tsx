import { memo, ReactElement } from "react";

import { Pronunciation } from "api/models";
import { getAudioUrl } from "backend";
import AudioPlayer from "components/Pronunciations/AudioPlayer";
import AudioRecorder from "components/Pronunciations/AudioRecorder";
import { FileWithSpeakerId } from "types/word";

interface PronunciationsBackendProps {
  audio: Pronunciation[];
  disabled?: boolean;
  playerOnly?: boolean;
  overrideMemo?: boolean;
  wordId: string;
  deleteAudio?: (fileName: string) => void;
  replaceAudio?: (audio: Pronunciation) => void;
  uploadAudio?: (file: FileWithSpeakerId) => void;
}

/** Audio recording/playing component for backend audio. */
export function PronunciationsBackend(
  props: PronunciationsBackendProps
): ReactElement {
  if (props.playerOnly && props.uploadAudio) {
    console.warn("uploadAudio is defined but unused since playerOnly is true");
  }
  if (!props.playerOnly && !props.uploadAudio) {
    console.warn("uploadAudio undefined; playerOnly should be set to true");
  }

  const audioButtons: ReactElement[] = props.audio.map((a) => (
    <AudioPlayer
      audio={a}
      deleteAudio={props.deleteAudio}
      disabled={props.disabled}
      key={a.fileName}
      pronunciationUrl={getAudioUrl(props.wordId, a.fileName)}
      updateAudioSpeaker={
        props.replaceAudio &&
        ((id) => props.replaceAudio!({ ...a, speakerId: id ?? "" }))
      }
    />
  ));

  return (
    <>
      {!props.playerOnly && !!props.uploadAudio && (
        <AudioRecorder
          disabled={props.disabled}
          id={props.wordId}
          uploadAudio={props.uploadAudio}
        />
      )}
      {audioButtons}
    </>
  );
}

// Memoize to decrease unnecessary fetching of audio files.
// https://dmitripavlutin.com/use-react-memo-wisely/#11-custom-equality-check-of-props
function propsAreEqual(
  prev: PronunciationsBackendProps,
  next: PronunciationsBackendProps
): boolean {
  if (next.overrideMemo) {
    return false;
  }
  return (
    prev.disabled === next.disabled &&
    prev.wordId === next.wordId &&
    JSON.stringify(prev.audio) === JSON.stringify(next.audio)
  );
}

export default memo(PronunciationsBackend, propsAreEqual);
