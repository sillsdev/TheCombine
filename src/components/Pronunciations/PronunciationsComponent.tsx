import React from "react";
import { LocalizeContextProps, withLocalize } from "react-localize-redux";
import AudioPlayer from "./AudioPlayer";
import AudioRecorder from "./AudioRecorder";
import * as Backend from "../../backend";
import theme from "../../types/theme";
import { Recorder } from "./Recorder";

export interface PronunciationProps {
  wordId: string;
  pronunciationFiles: string[];
  wordUpdated?: (oldId: string, newId: string) => void;
  recorder?: Recorder;
}

/** Audio recording/playing component */
export class Pronunciations extends React.Component<
  PronunciationProps & LocalizeContextProps
> {
  render() {
    let audioButtons;
    if (this.props.pronunciationFiles === undefined) {
      audioButtons = null;
    } else {
      audioButtons = this.props.pronunciationFiles.map((file) => {
        return (
          <AudioPlayer
            key={file}
            pronunciationUrl={Backend.getAudioUrl(this.props.wordId, file)}
          />
        );
      });
    }
    return (
      <div
        className="pronunciationAudio"
        style={{ paddingRight: theme.spacing(1) }}
      >
        <AudioRecorder
          key={this.props.wordId}
          wordId={this.props.wordId}
          recorder={this.props.recorder}
          recordingFinished={this.props.wordUpdated}
        />
        {audioButtons}
      </div>
    );
  }
}

export default withLocalize(Pronunciations);
