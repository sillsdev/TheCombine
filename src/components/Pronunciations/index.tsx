import { memo, ReactElement } from "react";

import { getAudioUrl } from "backend";
import AudioPlayer from "components/Pronunciations/AudioPlayer";
import AudioRecorder from "components/Pronunciations/AudioRecorder";

interface PronunciationProps {
  audioInFrontend?: boolean;
  pronunciationFiles: string[];
  overrideMemo?: boolean;
  spacer?: ReactElement;
  wordId: string;
  deleteAudio: (wordId: string, fileName: string) => void;
  uploadAudio?: (wordId: string, audioFile: File) => void;
}

/** Audio recording/playing component */
export function Pronunciations(props: PronunciationProps): ReactElement {
  const audioButtons: ReactElement[] = props.pronunciationFiles.map(
    (fileName) => (
      <AudioPlayer
        fileName={fileName}
        key={fileName}
        pronunciationUrl={
          props.audioInFrontend ? fileName : getAudioUrl(props.wordId, fileName)
        }
        deleteAudio={(fileName: string) =>
          props.deleteAudio(props.wordId, fileName)
        }
      />
    )
  );

  return (
    <>
      {!!props.uploadAudio && (
        <AudioRecorder wordId={props.wordId} uploadAudio={props.uploadAudio} />
      )}
      {props.spacer}
      {audioButtons}
    </>
  );
}

// Memoize to decrease unnecessary fetching of audio files.
// https://dmitripavlutin.com/use-react-memo-wisely/#11-custom-equality-check-of-props
function pronunciationPropsAreEqual(
  prev: PronunciationProps,
  next: PronunciationProps
): boolean {
  if (next.overrideMemo) {
    return false;
  }
  return (
    prev.wordId === next.wordId &&
    JSON.stringify(prev.pronunciationFiles) ===
      JSON.stringify(next.pronunciationFiles)
  );
}

export default memo(Pronunciations, pronunciationPropsAreEqual);
