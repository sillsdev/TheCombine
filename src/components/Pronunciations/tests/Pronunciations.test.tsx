import React from "react";
import ReactDOM from "react-dom";
import configureMockStore from "redux-mock-store";
import { defaultState } from "../../App/DefaultState";
import { Provider } from "react-redux";
import Pronunciations from "../PronunciationsComponent";
import AudioPlayer from "../AudioPlayer";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import AudioRecorder from "../AudioRecorder";

const createMockStore = configureMockStore([]);

// Mock the node module used by AudioRecorder
jest.mock("../Recorder");

// Variables
var testRenderer: ReactTestRenderer;

const mockStore = createMockStore(defaultState);

beforeAll(() => {
  renderer.act(() => {
    testRenderer = renderer.create(
      <Provider store={mockStore}>
        <Pronunciations wordId="2" pronunciationFiles={["a.wav", "b.wav"]} />
        />
      </Provider>
    );
  });
});

it("renders one record button and one play button for each pronunciation file", () => {
  expect(testRenderer.root.findAllByType(AudioRecorder).length).toBe(1);
  expect(testRenderer.root.findAllByType(AudioPlayer).length).toBe(2);
});

it("renders without crashing", () => {
  const mockStore = createMockStore(defaultState);
  const div = document.createElement("div");
  ReactDOM.render(
    <Provider store={mockStore}>
      <Pronunciations wordId="1" pronunciationFiles={[]} />
    </Provider>,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});
