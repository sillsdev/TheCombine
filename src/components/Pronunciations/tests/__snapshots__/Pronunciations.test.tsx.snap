// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pronunciation tests displays buttons 1`] = `
<div
  className="pronunciationAudio"
>
  <button
    aria-label="record"
    className="MuiButtonBase-root MuiIconButton-root makeStyles-button-6"
    disabled={false}
    id="recordingButton"
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-iconRelease-8"
        focusable="false"
        id="icon"
        viewBox="0 0 24 24"
      >
        <circle
          cx="12"
          cy="12"
          r="8"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
  <button
    aria-describedby={null}
    aria-label="play"
    className="MuiButtonBase-root MuiIconButton-root makeStyles-button-4"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    title={null}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-icon-5 MuiSvgIcon-colorPrimary"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M8 5v14l11-7z"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
  <button
    aria-describedby={null}
    aria-label="play"
    className="MuiButtonBase-root MuiIconButton-root makeStyles-button-4"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    title={null}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-icon-5 MuiSvgIcon-colorPrimary"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M8 5v14l11-7z"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
</div>
`;
